services/processor/native/repository/_Elections/processing_rewards.go

```diff
@@ -145,38 +144,48 @@ func getExcellenceProgramGuardians() []byte {
 	return state.ReadBytes(_formatExcellenceProgramGuardians())
 }
 
-func _setExcellenceProgramGuardians(guardians map[[20]byte]uint64) {
+func _setExcellenceProgramGuardians(guardians guardianArray) {
 	guardiansForSave := make([]byte, 0, len(guardians)*20)
-	for guardianAddr := range guardians {
-		guardiansForSave = append(guardiansForSave, guardianAddr[:]...)
+	for _, guardian := range guardians {
+		guardiansForSave = append(guardiansForSave, guardian.address[:]...)
 	}
 	state.WriteBytes(_formatExcellenceProgramGuardians(), guardiansForSave)
 }
 
 /***
  * Rewards: Sort top guardians using sort.Interface
  */
-func _getTopGuardians(guardiansAccumulatedStake map[[20]byte]uint64) (topGuardiansStake map[[20]byte]uint64, totalVotes uint64) {
+func _getTopGuardians(guardiansAccumulatedStake map[[20]byte]uint64) (topGuardiansStake guardianArray, totalVotes uint64) {
 	totalVotes = uint64(0)
-	topGuardiansStake = make(map[[20]byte]uint64)
 
 	guardianList := make(guardianArray, 0, len(guardiansAccumulatedStake))
 	for guardian, vote := range guardiansAccumulatedStake {
 		guardianList = append(guardianList, &guardianVote{guardian, vote})
 	}
 	sort.Sort(guardianList)
 
-	for i := 0; i < ELECTION_GUARDIAN_EXCELLENCE_MAX_NUMBER; i++ {
-		fmt.Printf("elections %10d rewards: top guardian %x, has %d votes\n", _getCurrentElectionBlockNumber(), guardianList[i].guardian, guardianList[i].vote)
+	i := 0
+	for i = 0; i < len(guardianList) && i < ELECTION_GUARDIAN_EXCELLENCE_MAX_NUMBER; i++ {
```

Comment:
if you replaced ` i < len(guardianList) && i < ELECTION_GUARDIAN_EXCELLENCE_MAX_NUMBER;` with ` i < len(guardianList) && isEligibleForReward(guardianList, i);` you can remove the second loop. 

`isEligibleForReward` can contain the current condition with an OR: `i > 0 && guardianList[i].vote != guardianList[i-1].vote && isEligibleForReward(guardianList[i-1])`

in words - a guardian is eligible if it's in the top `ELECTION_GUARDIAN_EXCELLENCE_MAX_NUMBER` OR if it's not the top guardian, and it has the same stake as the previous one, and the previous one was eligible.

Reply:
i've incorporated your idea in next version but cannot change it in this as its already deployed.